### ISSUE TITLE
immutable push* and append

### ISSUE THREAD

**Issue by exterm** [bodil/im-rs#40]
Hi!

I'm pretty new to rust, but playing around with it the first thing I got really annoyed with it was the lack of immutable lists. Recursion over mutable vectors is just no fun.

`im::Vector` seemed like the thing I wanted, but then I had really expected it to have

`pub fn append(&self, other: Self) -> Self`
`pub fn push_front(&self, value: A) -> Self`
etc.

Is it a conscious design decision to make these methods mutating instead?

**Comment by bodil** [comment]
Yes, that's very much a design decision - in fact, it used to work in the way you describe, but it ended up being very inconsistent with how you'd actually use these datatypes efficiently in Rust.

They're very much designed for use with recursive algorithms (in which case, `.clone()` will efficiently snapshot your current state once you're done mutating) but if you're looking for the Lisp/Haskell idiom of consing or unconsing using recursive functions, that's actually a terrible idea in Rust - it has no tail recursion, for a start! - and you should be using iterators and for loops instead, though it breaks my Haskeller's heart to say so.

**Comment by exterm** [comment]
It breaks my heart too :disappointed:  I got really enthusiastic after reading about traits and other aspects of the type system.

Thank you for your response! Is tail call optimization the only thing blocking this?

**Comment by bodil** [comment]
It's less about the features required to support it than about how Rust is designed overall with its focus on zero cost abstractions - recursive functions are just not a practical way to iterate over data structures at this level. I could imagine, at least given TCO, how to do certain operations over certain data structures (pretty much just array slices and cons lists) efficiently in Rust, but the data structures in this crate would perform very poorly with that pattern.

You can still write code that feels closer to Haskell that isn't terribly inefficient if you step your abstractions up a level and think of iterators like functors etc with some extra boilerplate, like:

```rust
let vec: Vector<_> = other_vec.into_iter().map(|a| Thing::from(a)).collect();
let vec: Vector<_> = im::iter::unfold(0, |n| (n, n + 1)).take(100).collect();
```

(The latter example is a bit contrived, because `Vector::from_iter(0..100)` would be a lot more sensible, but just to show that unfolding is a thing you can do.)

But still the awkward truth is that if you're mapping to the same type, this is orders of magnitude faster than collecting an iterator:

```rust
for ptr in vec.iter_mut() {
    *ptr += 1;
}
```
